pandas/tests/groupby/test_groupby.py

```diff
@@ -2348,6 +2348,41 @@ def test_groupby_duplicate_index():
     tm.assert_series_equal(result, expected)
 
 
+def test_group_on_empty_multiindex(transformation_func, request):
+    # GH 47787
+    # With one row, those are transforms so the schema should be the same
+    if transformation_func == "tshift":
+        mark = pytest.mark.xfail(raises=NotImplementedError)
+        request.node.add_marker(mark)
+    df = DataFrame(
+        data=[[1, Timestamp("today"), 3, 4]],
+        columns=["col_1", "col_2", "col_3", "col_4"],
+    )
+    df = df.set_index(["col_1", "col_2"])
+    if transformation_func == "fillna":
+        args = ("ffill",)
+    elif transformation_func == "tshift":
+        args = (1, "D")
+    else:
+        args = ()
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(df.index, result.index)
+
+    col_3 = df["col_3"]
+    result = col_3.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(col_3.index, result.index)
+
+    # When empty, expect the same schema as well
+    df = DataFrame(data=[], columns=["col_1", "col_2", "col_3", "col_4"], dtype=int)
+    df = df.set_index(["col_1", "col_2"])
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    assert df.index.names == result.index.names
```

Comment:
I think we should be testing using `tm.assert_frame_equal` and `tm.assert_series_equal` in all of these. We want to make sure we get the entire result right, not just the index names or the index.

Reply:
if that is the case,, we need to split because not all transformations are equals. The cumcount returns a single series and the fillna a dataframe. This felt like the right trade off as I am testing every transformation (and every future transformation will be picked up), but I can split in 2 tests if absolutely necessary.

Reply:
2 tests (or more!) is fine if one gets too complicated. Also can use `tm.assert_equal` to test on either Series or DataFrame. But we should always be testing the entire object.

Reply:
resubmitted, not sure why the docker test decided to fail